### ISSUE TITLE
compiletest interprets subdirectory for tests as filter instead

### ISSUE THREAD

**Issue by klensy** [rust-lang/rust#120300]
While working on https://github.com/rust-lang/rust/pull/120273, i used `python x.py --stage=1 t tests/ui/borrowck/ --force-rerun` to run some subset of tests instead of full suite, as described in https://rustc-dev-guide.rust-lang.org/tests/running.html#running-a-subset-of-the-test-suites:
```
If you only need to test a specific subdirectory of tests for any given test suite,
 you can pass that directory to ./x test:

./x test tests/ui/const-generics
```
and expected to run test only in tests/ui/borrowck/ folder,
but instead noticed that this actually tested not subdirectory `tests/ui/borrowck/`, but any test with `borrowck` in path, like
```
tests/ui/borrowck/borrowck-mut-addr-of-imm-var.rs
tests/ui/pattern/bindings-after-at/borrowck-pat-by-copy-bindings-in-at.rs
tests/ui/span/borrowck-ref-into-rvalue.rs
```

Or maybe bootstrap parses input for compiletest wrong?

@rustbot label +A-testsuite

**Comment by ehuss** [comment]
I believe this is a limitation of libtest, which is only capable of filtering tests based on "contains" or "exact", but not "startswith".

**Comment by klensy** [comment]
> I believe this is a limitation of libtest, which is only capable of filtering tests based on "contains" or "exact", but not "startswith".

Yes, but `compiletest` can filter/mark as ignored tests before passing to `test::run_tests_console`
